File: go-server/handlers/rooms.go
package handlers

import (
	"encoding/json"
	"net/http"
	"regexp"
	"strconv"
	"sync"
	"time"

	"github.com/XingLingQAQ/NextClip/go-server/middleware"
	"github.com/XingLingQAQ/NextClip/go-server/storage"
	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
)

var validExpiry = map[string]bool{"1": true, "24": true, "168": true, "720": true, "permanent": true}
var pinRegex = regexp.MustCompile(`^\d{6}$`)

const roomTokenTTL = 24 * time.Hour

type roomTokenEntry struct {
	expiresAt time.Time
}

// RoomTokenManager manages room access tokens.
type RoomTokenManager struct {
	mu     sync.RWMutex
	tokens map[string]map[string]*roomTokenEntry // roomCode -> token -> entry
}

func NewRoomTokenManager() *RoomTokenManager {
	rtm := &RoomTokenManager{tokens: make(map[string]map[string]*roomTokenEntry)}
	go rtm.cleanupLoop()
	return rtm
}

func (rtm *RoomTokenManager) cleanupLoop() {
	ticker := time.NewTicker(60 * time.Second)
	for range ticker.C {
		rtm.mu.Lock()
		now := time.Now()
		for roomCode, tokens := range rtm.tokens {
			for token, entry := range tokens {
				if entry.expiresAt.Before(now) {
					delete(tokens, token)
				}
			}
			if len(tokens) == 0 {
				delete(rtm.tokens, roomCode)
			}
		}
		rtm.mu.Unlock()
	}
}

func (rtm *RoomTokenManager) Issue(roomCode string) string {
	token := uuid.New().String()
	rtm.mu.Lock()
	defer rtm.mu.Unlock()

	if _, ok := rtm.tokens[roomCode]; !ok {
		rtm.tokens[roomCode] = make(map[string]*roomTokenEntry)
	}
	rtm.tokens[roomCode][token] = &roomTokenEntry{expiresAt: time.Now().Add(roomTokenTTL)}
	return token
}

func (rtm *RoomTokenManager) Validate(roomCode, token string) bool {
	rtm.mu.RLock()
	defer rtm.mu.RUnlock()

	roomTokens, ok := rtm.tokens[roomCode]
	if !ok {
		return false
	}
	entry, ok := roomTokens[token]
	if !ok {
		return false
	}
	if entry.expiresAt.Before(time.Now()) {
		return false
	}
	return true
}

func (rtm *RoomTokenManager) Revoke(roomCode string) {
	rtm.mu.Lock()
	defer rtm.mu.Unlock()
	delete(rtm.tokens, roomCode)
}

// RoomHandler handles room-related API endpoints.
type RoomHandler struct {
	store    *storage.Store
	tokenMgr *RoomTokenManager
	rl       *middleware.RateLimiter
}

func NewRoomHandler(store *storage.Store, tokenMgr *RoomTokenManager, rl *middleware.RateLimiter) *RoomHandler {
	return &RoomHandler{store: store, tokenMgr: tokenMgr, rl: rl}
}

func (h *RoomHandler) GetRoom(w http.ResponseWriter, r *http.Request) {
	roomCode := chi.URLParam(r, "roomCode")
	room, err := h.store.GetRoom(roomCode)
	if err != nil || room == nil {
		writeJSON(w, http.StatusOK, map[string]interface{}{"exists": false})
		return
	}

	user := middleware.GetCurrentUser(r)
	token := getRoomTokenFromRequest(r)
	canManage := user != nil && room.OwnerID != nil && *room.OwnerID == user.ID && token != "" && h.tokenMgr.Validate(room.RoomCode, token)

	resp := map[string]interface{}{
		"exists":      true,
		"hasPassword": room.HasPassword,
		"canManage":   canManage,
	}
	if canManage {
		resp["expiresAt"] = room.ExpiresAt
		resp["createdAt"] = room.CreatedAt
	}
	writeJSON(w, http.StatusOK, resp)
}

func (h *RoomHandler) JoinRoom(w http.ResponseWriter, r *http.Request) {
	roomCode := chi.URLParam(r, "roomCode")

	var body struct {
		Password string `json:"password"`
	}
	json.NewDecoder(r.Body).Decode(&body)

	room, _ := h.store.GetRoom(roomCode)

	if room == nil {
		// Create room with 24h expiry
		expiresAt := time.Now().Add(24 * time.Hour).UTC().Format(time.RFC3339Nano)
		user := middleware.GetCurrentUser(r)
		var ownerID *string
		if user != nil {
			ownerID = &user.ID
		}
		h.store.CreateRoom(roomCode, ownerID, nil, &expiresAt)
		token := h.tokenMgr.Issue(roomCode)
		writeJSON(w, http.StatusOK, map[string]interface{}{
			"success":     true,
			"created":     true,
			"hasPassword": false,
			"token":       token,
		})
		return
	}

	if room.HasPassword {
		if body.Password == "" {
			writeJSON(w, http.StatusUnauthorized, map[string]interface{}{
				"message":      "Password required",
				"needPassword": true,
			})
			return
		}
		if !h.store.VerifyRoomPassword(roomCode, body.Password) {
			writeJSON(w, http.StatusForbidden, map[string]string{"message": "Incorrect password"})
			return
		}
	}

	token := h.tokenMgr.Issue(roomCode)
	writeJSON(w, http.StatusOK, map[string]interface{}{
		"success":     true,
		"created":     false,
		"hasPassword": room.HasPassword,
		"expiresAt":   room.ExpiresAt,
		"token":       token,
	})
}

func (h *RoomHandler) SetPassword(w http.ResponseWriter, r *http.Request) {
	roomCode := chi.URLParam(r, "roomCode")
	user := middleware.GetCurrentUser(r)
	if user == nil {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"message": "Unauthorized"})
		return
	}

	var body struct {
		Password string `json:"password"`
		Token    string `json:"token"`
	}
	json.NewDecoder(r.Body).Decode(&body)

	if !h.tokenMgr.Validate(roomCode, body.Token) {
		writeJSON(w, http.StatusForbidden, map[string]string{"message": "Unauthorized"})
		return
	}

	if !h.assertRoomOwner(w, roomCode, user.ID, true) {
		return
	}

	if body.Password != "" {
		if !pinRegex.MatchString(body.Password) {
			writeJSON(w, http.StatusBadRequest, map[string]string{"message": "Password must be exactly 6 digits"})
			return
		}
	}

	var pass *string
	if body.Password != "" {
		pass = &body.Password
	}
	h.store.SetRoomPassword(roomCode, pass)
	h.tokenMgr.Revoke(roomCode)
	writeJSON(w, http.StatusOK, map[string]interface{}{"success": true})
}

func (h *RoomHandler) SetExpiry(w http.ResponseWriter, r *http.Request) {
	roomCode := chi.URLParam(r, "roomCode")
	user := middleware.GetCurrentUser(r)
	if user == nil {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"message": "Unauthorized"})
		return
	}

	var body struct {
		ExpiryHours interface{} `json:"expiryHours"`
		Token       string      `json:"token"`
	}
	json.NewDecoder(r.Body).Decode(&body)

	if !h.tokenMgr.Validate(roomCode, body.Token) {
		writeJSON(w, http.StatusForbidden, map[string]string{"message": "Unauthorized"})
		return
	}

	if !h.assertRoomOwner(w, roomCode, user.ID, true) {
		return
	}

	var expiryStr string
	switch v := body.ExpiryHours.(type) {
	case string:
		expiryStr = v
	case float64:
		expiryStr = strconv.FormatFloat(v, 'f', -1, 64)
	default:
		expiryStr = ""
	}

	if !validExpiry[expiryStr] {
		writeJSON(w, http.StatusBadRequest, map[string]string{"message": "Invalid expiry value"})
		return
	}

	var expiresAt *string
	if expiryStr != "permanent" {
		hours, _ := strconv.Atoi(expiryStr)
		exp := time.Now().Add(time.Duration(hours) * time.Hour).UTC().Format(time.RFC3339Nano)
		expiresAt = &exp
	}

	h.store.SetRoomExpiry(roomCode, expiresAt)
	writeJSON(w, http.StatusOK, map[string]interface{}{"success": true, "expiresAt": expiresAt})
}

func (h *RoomHandler) assertRoomOwner(w http.ResponseWriter, roomCode, userID string, hasValidToken bool) bool {
	room, _ := h.store.GetRoom(roomCode)
	if room == nil {
		writeJSON(w, http.StatusNotFound, map[string]string{"message": "Room not found"})
		return false
	}
	if room.OwnerID == nil {
		if !hasValidToken {
			writeJSON(w, http.StatusForbidden, map[string]string{"message": "Only the room owner can perform this action"})
			return false
		}
		return true
	}
	if *room.OwnerID != userID {
		writeJSON(w, http.StatusForbidden, map[string]string{"message": "Only the room owner can perform this action"})
		return false
	}
	return true
}

File: worker/src/rooms.ts
import { Hono } from "hono";
import type { Env, User } from "./types";
import { hashPassword, verifyPassword } from "./crypto";
import { issueRoomToken, validateRoomToken, revokeRoomTokens } from "./token-store";

// Re-export for use in clips.ts
export { validateRoomToken };

const rooms = new Hono<{ Bindings: Env; Variables: { user?: User } }>();

// Get room info
rooms.get("/:roomCode", async (c) => {
  const roomCode = c.req.param("roomCode");
  const row = await c.env.DB.prepare(
    "SELECT room_code, password_hash, owner_id, expires_at, created_at FROM rooms WHERE room_code = ?"
  ).bind(roomCode).first<any>();
  if (!row) return c.json({ exists: false });

  const user = c.get("user");
  const token = c.req.header("X-Room-Token") || c.req.query("token") || "";
  let canManage = false;
  if (user && row.owner_id === user.id && token) {
    canManage = await validateRoomToken(c.env.DB, roomCode, token);
  }

  const resp: any = { exists: true, hasPassword: !!row.password_hash, canManage };
  if (canManage) { resp.expiresAt = row.expires_at; resp.createdAt = row.created_at; }
  return c.json(resp);
});

// Join room
rooms.post("/:roomCode/join", async (c) => {
  const roomCode = c.req.param("roomCode");
  const body: { password?: string } = await c.req.json<{ password?: string }>().catch(() => ({}));

  const row = await c.env.DB.prepare(
    "SELECT room_code, password_hash, owner_id, expires_at FROM rooms WHERE room_code = ?"
  ).bind(roomCode).first<any>();

  if (!row) {
    const expiresAt = new Date(Date.now() + 24 * 3600000).toISOString();
    const user = c.get("user");
    await c.env.DB.prepare(
      "INSERT INTO rooms (room_code, password_hash, owner_id, expires_at, created_at) VALUES (?, ?, ?, ?, ?)"
    ).bind(roomCode, null, user?.id || null, expiresAt, new Date().toISOString()).run();
    const token = await issueRoomToken(c.env.DB, roomCode);
    return c.json({ success: true, created: true, hasPassword: false, token });
  }

  if (row.password_hash) {
    if (!body.password) return c.json({ message: "Password required", needPassword: true }, 401);
    const result = await verifyPassword(body.password, row.password_hash);
    if (!result.matched) return c.json({ message: "Incorrect password" }, 403);
    if (result.needsRehash) {
      const newHash = await hashPassword(body.password);
      await c.env.DB.prepare("UPDATE rooms SET password_hash = ? WHERE room_code = ?").bind(newHash, roomCode).run();
    }
  }

  const token = await issueRoomToken(c.env.DB, roomCode);
  return c.json({ success: true, created: false, hasPassword: !!row.password_hash, expiresAt: row.expires_at, token });
});

// Set password
rooms.post("/:roomCode/password", async (c) => {
  const user = c.get("user");
  if (!user) return c.json({ message: "Unauthorized" }, 401);
  const roomCode = c.req.param("roomCode");
  const body = await c.req.json<{ password?: string; token?: string }>();
  if (!body.token || !(await validateRoomToken(c.env.DB, roomCode, body.token))) {
    return c.json({ message: "Unauthorized" }, 403);
  }

  if (body.password && !/^\d{6}$/.test(body.password)) {
    return c.json({ message: "Password must be exactly 6 digits" }, 400);
  }

  const hash = body.password ? await hashPassword(body.password) : null;
  await c.env.DB.prepare("UPDATE rooms SET password_hash = ? WHERE room_code = ?").bind(hash, roomCode).run();
  await revokeRoomTokens(c.env.DB, roomCode);
  return c.json({ success: true });
});

// Set expiry
rooms.post("/:roomCode/expiry", async (c) => {
  const user = c.get("user");
  if (!user) return c.json({ message: "Unauthorized" }, 401);
  const roomCode = c.req.param("roomCode");
  const body = await c.req.json<{ expiryHours?: string | number; token?: string }>();
  if (!body.token || !(await validateRoomToken(c.env.DB, roomCode, body.token))) {
    return c.json({ message: "Unauthorized" }, 403);
  }

  const validExpiry = new Set(["1", "24", "168", "720", "permanent"]);
  const expiryStr = String(body.expiryHours || "");
  if (!validExpiry.has(expiryStr)) return c.json({ message: "Invalid expiry value" }, 400);

  let expiresAt: string | null = null;
  if (expiryStr !== "permanent") {
    expiresAt = new Date(Date.now() + parseInt(expiryStr) * 3600000).toISOString();
  }
  await c.env.DB.prepare("UPDATE rooms SET expires_at = ? WHERE room_code = ?").bind(expiresAt, roomCode).run();
  return c.json({ success: true, expiresAt });
});

export default rooms;
